public/jira/rollup/dates/dates.ts

```diff
@@ -0,0 +1,213 @@
+
+const methods = {
+  parentFirstThenChildren,
+  childrenOnly,
+  childrenFirstThenParent,
+  widestRange,
+  parentOnly,
+};
+export type CalculationName = keyof typeof methods;
+
+export type RollupDateData = Partial<StartData & DueData>;
+
+export type RolledupDatesReleaseOrIssue = IssueOrRelease<DerivedIssue> & {
+  rollupDates: RollupDateData;
+};
+
+const getStartData = (d: Partial<StartData> | null) => {
+  if(!d)return {}
+  
+  const {start,startFrom}=d
+
+  return ({
+    ...(start && { start }),
+    ...(startFrom && { startFrom }),
+  });
+};
+const getDueData = (d: Partial<DueData> | null) => {
+  if(!d)return {}
+  
+  const {due,dueTo}=d
+
+  return ({
+    ...(due && { due }),
+    ...(dueTo && { dueTo }),
+  });
+};
+
+/**
+ *
+ * @param {Array<IssueOrRelease>} issuesOrReleases Starting from low to high
+ * @param {Array<String>} methodNames Starting from low to high
+ * @return {Array<RollupDateData>}
+ */
+export function rollupDates<_CustomFields, Meta>(
+  groupedHierarchy: DerivedIssue[][],
+  methodNames: CalculationName[]
+): RollupResponse<RollupDateData, Meta> {
+  return rollupGroupedHierarchy(groupedHierarchy, {
+    createRollupDataFromParentAndChild(
+      issueOrRelease: ReportingHierarchyIssueOrRelease<DerivedIssue>,
+      children: RollupDateData[],
+      hierarchyLevel
+    ) {
+      const methodName =
+        methodNames[hierarchyLevel] || "childrenFirstThenParent";
+      const method = methods[methodName];
+      const result = method(issueOrRelease, children);
+      return result;
+    },
+  });
+}
+
+/**
+ *
+ * @param {Array<IssueOrRelease>} issuesOrReleases
+ * @param {Array<{type: String, hierarchyLevel: Number, calculation: String}>} rollupTimingLevelsAndCalculations
+ * @return {Array<RolledupDatesReleaseOrIssue>}
+ */
+export function addRollupDates(
+  issuesOrReleases: DerivedIssue[],
+  rollupTimingLevelsAndCalculations: {
+    type: string;
+    hierarchyLevel: number;
+    calculation: CalculationName;
+  }[]
+) {
+  const groupedIssues = groupIssuesByHierarchyLevelOrType(
+    issuesOrReleases,
+    rollupTimingLevelsAndCalculations
+  );
+  const rollupMethods = rollupTimingLevelsAndCalculations
+    .map((rollupData) => rollupData.calculation)
+    .reverse();
+  const rolledUpDates = rollupDates(groupedIssues, rollupMethods);
+  const zipped = zipRollupDataOntoGroupedData(
+    groupedIssues,
+    rolledUpDates,
+    "rollupDates"
+  );
+  return zipped.flat();
+}
+
+export function mergeStartAndDueData<T extends Partial<StartData & DueData>>(
+  records: T[]
+) {
+  const startData = records.filter(selectStart).map(getStartData);
+  const dueData = records.filter(selectDue).map(getDueData);
+
+  const sortedStart = startData.sort(sortByStart);
+  const sortedDue = dueData.sort(sortByDue);
+
+  const sd = startData.length ? sortedStart[0] : {};
+  const ed = dueData.length ? sortedDue[0] : {};
+
+  return {
+    ...sd,
+    ...ed,
+  } as T;
```

Comment:
Try defining the return type to be `T`  instead of casting for inference. If there's a disconnect there we need to understand what it is before we decide to cast